### PR TITLE
DATAUP-737: more Deepscan fixes for global variables and function params

### PR DESCRIPTION
**Pull request by ialarmedalien**
# Description of PR purpose/changes

This turned into a bit of a horror show, trying to fix some of the errors that deep scan revealed.

What I did:
- Looked through the [Deepscan issues from the truss => develop merge](https://deepscan.io/dashboard/#view=project&tid=17407&pid=20771&bid=575925&prid=1192118&subview=issues)
- Fixed all the issues that could be easily located and required minimal brain effort
- Used the eslint and sonarcloud extensions in VSCode to automatically fix (some of the) other issues in those files
- If the module did not `'use strict'`, turned it on; if there were issues that would crash the widget (e.g. missing dependencies or global / undeclared variables), I turned it off again.
- Wished numerous times that I had not started this effort!

# Jira Ticket / Issue #

Related Jira ticket: https://kbase-jira.atlassian.net/browse/DATAUP-737
- [x] Added the Jira Ticket to the title of the PR (e.g. `DATAUP-69 Adds a PR template`)

# Testing Instructions
* Details for how to test the PR:
- [x] Tests pass locally and in GitHub Actions
- [ ] Changes available by spinning up a local narrative and navigating to _X_ to see _Y_

# Dev Checklist:

- [x] My code follows the guidelines at https://sites.google.com/lbl.gov/trussresources/home?authuser=0
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] (JavaScript) I have run Prettier and ESLint on changed code manually or with a git precommit hook

### CODE REVIEW

File: kbase-extension/static/kbase/js/common/iframe/messages.js [hunk 1]
```diff
@@ -30,7 +30,7 @@ define(['bluebird', 'uuid'], (Promise, Uuid) => {
             const message = event.data;
             let response;
 
-            if (!message.address && !message.address.to) {
+            if (!message.address || !message.address.to) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/iframe/messages.js:33]: deepscan spotted this error

File: kbase-extension/static/kbase/js/common/parameterSpec.js [hunk 2]
```diff
@@ -150,7 +147,7 @@ define([], () => {
                 case 'float':
                     return parseFloat(_defaultValue);
                 case 'boolean':
-                    return coerceToBoolean(_defaultValue);
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/parameterSpec.js:153]: method does not exist

File: kbase-extension/static/kbase/js/common/parameterSpec.js [hunk 3]
```diff
@@ -43,10 +43,7 @@ define([], () => {
         }
 
         function isAdvanced() {
-            if (spec.advanced === 1) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/parameterSpec.js:46]: sonarcloud optimisation

File: kbase-extension/static/kbase/js/common/parameterSpec.js [hunk 4]
```diff
@@ -177,7 +174,7 @@ define([], () => {
             if (typeof value !== 'string') {
                 return 0;
             }
-            switch (value.toLowerCase(value)) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/parameterSpec.js:180]: `toLowerCase()` should not have a parameter

File: kbase-extension/static/kbase/js/common/parameterSpec.js [hunk 5]
```diff
@@ -237,13 +234,10 @@ define([], () => {
             if (value === undefined || value === null) {
                 return true;
             }
-            if (
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/parameterSpec.js:240]: sonar cloud optimisation

File: kbase-extension/static/kbase/js/common/sdk.js [hunk 6]
```diff
@@ -15,7 +15,7 @@ define(['common/props'], (Props) => {
         if (typeof value !== 'string') {
             return false;
         }
-        switch (value.toLowerCase(value)) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/sdk.js:18]: no param

File: kbase-extension/static/kbase/js/common/sdk.js [hunk 7]
```diff
@@ -578,7 +577,7 @@ define(['common/props'], (Props) => {
         if (spec.allow_multiple) {
             // except, ahem, for the custom_subdata, at least for now...
             if (spec.field_type === 'custom_textsubdata') {
-                spec.allow_multiple === 0;
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/sdk.js:581]: `===` ???

File: kbase-extension/static/kbase/js/common/sdk.js [hunk 8]
```diff
@@ -770,10 +765,7 @@ define(['common/props'], (Props) => {
         // and then add the groups in.
         const parameterLayout = sdkAppSpec.parameters
             .filter((parameter) => {
-                if (parameterSpecs[parameter.id]) {
-                    return true;
-                }
-                return false;
+                return !!parameterSpecs[parameter.id];
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/sdk.js:768]: sonarcloud optimisation

File: kbase-extension/static/kbase/js/common/sdk.js [hunk 9]
```diff
@@ -786,10 +778,7 @@ define(['common/props'], (Props) => {
                     // first filter out any groups which were not added to the parameters.
                     // This includes ones with no parameters specified
                     .filter((group) => {
-                        if (parameterSpecs[group.id]) {
-                            return true;
-                        }
-                        return false;
+                        return !!parameterSpecs[group.id];
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/sdk.js:781]: sonarcloud optimisation

File: kbase-extension/static/kbase/js/common/ui.js [hunk 10]
```diff
@@ -1193,7 +1193,7 @@ define([
                     fontSize = jsonBlockWidgetConfig.fontSize || 0.8;
 
                 function render(obj) {
-                    const specText = JSON.stringify(obj, false, indent),
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/ui.js:1196]: invalid second argument -- should be `null` or `undefined` if there is no transformation function

File: kbase-extension/static/kbase/js/common/viewModel/sdkImport.js [hunk 11]
```diff
@@ -521,46 +430,6 @@ define(['common/props'], (Props) => {
         };
     }
 
-    function convertGroupList(group, params) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/viewModel/sdkImport.js:524]: unused

File: kbase-extension/static/kbase/js/widgets/appWidgets2/fieldWidget.js [hunk 12]
```diff
@@ -145,12 +142,6 @@ define([
             places.$messagePanel.addClass('hidden');
         }
 
-        function hideError() {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/appWidgets2/fieldWidget.js:148]: unused

File: kbase-extension/static/kbase/js/widgets/appWidgets2/parameterSpec.js [hunk 13]
```diff
@@ -189,7 +167,7 @@ define([], () => {
                 case 'workspaceObjectName':
                     return defaultValue;
                 case 'boolean':
-                    return coerceToBoolean(defaultValue);
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/appWidgets2/parameterSpec.js:192]: method does not exist

File: kbase-extension/static/kbase/js/widgets/appWidgets2/parameterSpec.js [hunk 14]
```diff
@@ -234,23 +212,17 @@ define([], () => {
...
-                     * default value validator?
-                     */
-                    if (!defaultValues || defaultValues.length === 0) {
-                        return spec.checkbox_options.unchecked_value;
-                    } else {
-                        return coerceToIntBoolean(defaultValues[0]);
-                    }
-                case 'custom_textsubdata':
-                    if (!defaultValues) {
-                    }
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/appWidgets2/parameterSpec.js:253]: code does nothing

File: kbase-extension/static/kbase/js/widgets/function_input/kbaseDefaultNarrativeInput.js [hunk 15]
```diff
@@ -135,18 +136,16 @@ define(['kbwidget', 'bootstrap', 'jquery', 'narrativeConfig', 'kbaseNarrativeInp
...
                     const fieldName = $field.attr('name');
 
                     // If it's a text field, just dump the value in there.
-                    if ($field.is('input') && $field.attr('type') === 'text') {
-                        $field.val(state[fieldName]);
-                    }
-
-                    // If it's a select field, do the same... we'll have comboboxen or something,
-                    // eventually, so I'm just leaving this open for that.
-                    else if ($field.is('select')) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/function_input/kbaseDefaultNarrativeInput.js:149]: these two different options do the same thing

File: test/unit/spec/narrative_core/upload/stagingAreaViewer-spec.js [hunk 16]
```diff
@@ -432,8 +432,7 @@ define([
                 bulkImportType = 'sra_reads',
                 bulkOutputSuffix = '_reads',
                 nonBulkAppId = 'kb_uploadmethods/import_tsv_as_expression_matrix_from_staging',
-                nonBulkType = 'expression_matrix',
-                nonBulkSuffix = '_matrix';
```
Comment (ialarmedalien) [inline test/unit/spec/narrative_core/upload/stagingAreaViewer-spec.js:436]: not used

File: nbextensions/appCell2/widgets/appCellWidget.js [hunk 17]
```diff
@@ -1427,7 +1427,7 @@ define(
                     newJobState = message[existingJobId].jobState,
                     { outputWidgetInfo } = message[existingJobId],
                     forceRender =
-                        !Jobs.isValidJobStateObject(existingState) &&
+                        (!existingState || !Jobs.isValidJobStateObject(existingState)) &&
```
Comment (ialarmedalien) [inline nbextensions/appCell2/widgets/appCellWidget.js:1430]: check that `existingState` is non-null

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js [hunk 18]
```diff
@@ -133,7 +131,7 @@ define([
                 .then((data) => {
                     const files = JSON.parse(data).map((f) => {
                         f.path = f.path.substring(f.path.indexOf('/') + 1);
-                        f.subdir = '/' + f.path.substring(0, f.path.lastIndexOf('/')) || '/';
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:136]: the `|| '/'` never gets used as the previous expression always evaluates to true

File: kbase-extension/static/kbase/js/widgets/kbaseButtonControls.js [hunk 19]
```diff
@@ -113,9 +114,6 @@ define([
             this._rewireIds($elem, this);
 
             if (this.options.onMouseover && this.options.type == 'floating') {
-                const overControls = false;
-                let overParent = false;
-
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/kbaseButtonControls.js:118]: removed a load of unused variables from this file

File: kbase-extension/static/kbase/js/widgets/kbStandalonePlot.js [hunk 20]
```diff
@@ -458,9 +445,6 @@
                     renderer.settings.title_color,
                     renderer.settings.title_settings
                 );
-            for (i = 0; i < renderer.settings.data.length; i++) {
-                const d = renderer.settings.data[i];
-            }
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/kbStandalonePlot.js:463]: iteration does nothing

File: kbase-extension/static/kbase/js/widgets/kbStandaloneGraph.js [hunk 21]
```diff
@@ -554,16 +542,16 @@
                 svg.graph.noDraw().gridlines({ stroke: 'gray', strokeDashArray: '2,2' }, 'gray');
             }
 
-            for (i = 0; i < renderer.settings.data.length; i++) {
+            for (const element of renderer.settings.data) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/kbStandaloneGraph.js:545]: thank you sonar cloud!

File: kbase-extension/static/kbase/js/widgets/function_output/modeling/KBaseFBA.FBAComparison.js [hunk 22]
```diff
@@ -462,123 +458,108 @@ function KBaseFBA_FBAComparison(modeltabs) {
...
-            this.modelcompounds[i].dispid = idarray[0] + '[' + idarray[1] + ']';
-            this.modelcompounds[i].exchange =
-                ' => ' + this.modelcompounds[i].name + '[' + namearray[1] + ']';
-            this.modelcompounds[i].numfba = 0;
-            var percent = Math.floor(
-                100 *
-                    this.modelcompounds[i].state_conservation[
-                        this.modelcompounds[i].most_common_state
-                    ][1]
+        for (const element of this.modelcompounds) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/function_output/modeling/KBaseFBA.FBAComparison.js:520]: refactor by sonar cloud

File: kbase-extension/static/kbase/js/widgets/appWidgets2/fieldWidgetCompact.js [hunk 23]
```diff
@@ -189,50 +188,50 @@ define([
             places.feedbackIndicator.className = 'kb-app-parameter-right-error-bar';
         }
 
-        function rawSpec(_spec) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/appWidgets2/fieldWidgetCompact.js:192]: use the existing `spec` variable

File: kbase-extension/static/kbase/js/widgets/appWidgets2/fieldWidgetBare.js [hunk 24]
```diff
@@ -206,21 +193,21 @@ define([
             }
         }
 
-        function rawSpec(spec) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/appWidgets2/fieldWidgetBare.js:209]: spec already defined in the upper scope

File: kbase-extension/static/kbase/js/common/parameterSpec.js (hunk 2 above)
Comment (briehl) [reply kbase-extension/static/kbase/js/common/parameterSpec.js:153]: That's wild. Maybe in a util module?
Maybe this one doesn't even get used? I wonder if it's cruft.

File: kbase-extension/static/kbase/js/common/parameterSpec.js (hunk 5 above)
Comment (briehl) [reply kbase-extension/static/kbase/js/common/parameterSpec.js:240]: The `!!` seems a little redundant. Shouldn't that whole expression resolve to a boolean? I guess it makes it a little easier to see at a glance.

File: kbase-extension/static/kbase/js/common/viewModel/sdkImport.js [hunk 25]
```diff
@@ -609,13 +478,11 @@ define(['common/props'], (Props) => {
...
     }
 
-    function makeListSpec(spec, params) {}
-
     function convertGroup(group, params) {
         const structSpec = convertGroupToStruct(group, params);
         if (group.allow_multiple === 1) {
             // wrap in a list
-            console.log('GROUP', group);
+            // console.log('GROUP', group);
```
Comment (briehl) [inline kbase-extension/static/kbase/js/common/viewModel/sdkImport.js:485]: Could just delete that. It doesn't seem important.

**Review by briehl — APPROVED** [review]
This all seems fine. I ran it locally, and everything seems to check out ok, too.

**Pull request merged**